### PR TITLE
Fixes ActivityIndicatorIOS doesn't hide initially

### PR DESCRIPTION
**Pull request by redmar**
I used the reproducible steps as described in origin bug ticket #7987 as test plan.
This has the same contents as PR #8130 but then against master per @janicduplessis request.

### CODE REVIEW

**Comment by facebook-github-bot-4** [comment]
Thank you for your pull request and welcome to our community. We require contributors to sign our Contributor License Agreement, and we don't seem to have you on file. In order for us to review and merge your code, please sign up at https://code.facebook.com/cla - and if you have received this in error or have any questions, please drop us a line at cla@fb.com. Thanks!

**Comment by ide** [comment]
Can you explain the root cause of the bug?

**Comment by facebook-github-bot-9** [comment]
Thank you for signing our Contributor License Agreement. We can now accept your code for this (and any) Facebook open source project. Thanks!

**Comment by redmar** [comment]
> Can you explain the root cause of the bug?

No not really 😞 , I've tried to trace diffs that relate to the ActivityIndicator but couldn't find the real origin. 
I think @janicduplessis comment and idea how to fix in #7987 could be a pointer but am not deep enough into the core of iOS react native to give any sensible reason why the bug occurred.
Still curious about the root cause.

**Comment by janicduplessis** [comment]
I know that calling stopRefreshing too early doesn't work, it's probably a but in UIKit, not sure exactly at what  point does it start working but calling if after the first layout does the trick. My only concern is that I think it used to work at some point so I wonder what changed to break this.

**Comment by janicduplessis** [comment]
@redmar Do you know if it is broken since a specific version or RN or did it never actually work?

**Comment by redmar** [comment]
@janicduplessis ok so this took me some time but I did a manual bisect of all the commits between 0.26 (good) and 0.27.0-rc (bad). I can pinpoint the bug to one specific commit now, even to 3 lines in that commit that enable it:

The commit where the bug is introduced is: fe5c0d2d0696b4fc5cdd65f1f2198c4f4363e543
(link to it: https://github.com/facebook/react-native/commit/fe5c0d2d0696b4fc5cdd65f1f2198c4f4363e543 )

and the 'offending' lines that enable the visibility of the ActivityIndicator while it should be hidden are:
https://github.com/facebook/react-native/blob/fe5c0d2d0696b4fc5cdd65f1f2198c4f4363e543/React/Modules/RCTUIManager.m#L643-L645

when i comment those lines out everything is working fine.

**Comment by janicduplessis** [comment]
Great job finding out the commit that broke this!

@rigdern Any idea on the best way to fix this?

**Comment by ide** [comment]
@redmar thanks for tracking that down. I think what is happening is that when the ActivityIndicator is stopped (which is its default state), it also internally sets `self.hidden = NO` unless you tell it not to. But this behavior is not modeled in React nor the shadow views so we end up doing:
```objc
actualActivityIndicator.hidden = shadowView.hidden // YES
```
These are three possible solutions:
1. Extend UIActivityIndicatorView and override `setHidden` to check `hidesWhenStopped` and `isAnimating`. This would make it so that we could never display a stopped ActivityIndicator, which on one hand is not very useful, but on the other, it is kind of lame to take away that flexibility unless we have to.

2. Make `shadowView.isHidden` (https://github.com/facebook/react-native/blob/fe5c0d2d0696b4fc5cdd65f1f2198c4f4363e543/React/Modules/RCTUIManager.m#L571) a tri-state NSValue boolean: YES, NO, nil. Set `view.hidden = isHidden` only when `isHidden != nil`. This does add complexity (boo) and maybe perf overhead (boo, but need to profile).

3. Break the implicit `hidesWhenStopped` behavior and make it so that if you want to hide the ActivityIndicator, you need to explicitly write `<ActivityIndicator hidden />` in React.

Maybe there are more ideas but those are the ones I have.

**Comment by redmar** [comment]
@ide I think I got your first point working but without the catch of not being able to never display a stopped ActivityIndicator. See added commit.

Tested this on the test case and it seems to work 😄

**Comment by janicduplessis** [comment]
@redmar What if `hidesWhenStopped` is set to false, does it still work?

**Comment by redmar** [comment]
@janicduplessis 
```
hidesWhenStopped={false} animating={false} => activity indicator initially shown but not animating
hidesWhenStopped={false} animating={true}  => activity indicator initially shown and animating
hidesWhenStopped={true} animating={true}  => activity indicator initially shown and animating
hidesWhenStopped={true} animating={false}  => activity indicator initially not shown (bug case, now working as expected)
```
AFAIK it all works as expected in these cases.

Current patched behaviour is now that when the 'offending' code (seen here https://github.com/facebook/react-native/blob/fe5c0d2d0696b4fc5cdd65f1f2198c4f4363e543/React/Modules/RCTUIManager.m#L643-L645 )  wants to setHidden=NO on the activityindicator view because the shadowview didn't get the memo to be hidden (as @ide described) it checks again if it should be hidden or not. It has logic to watch for when hidesWhenStopped=TRUE and animating=FALSE it should be hidden no matter what. That's why it's now hidden in the bug case (expected behaviour).

All other interaction with the control is delegated to super so should still work as intended.

**Comment by janicduplessis** [comment]
@redmar @ide What about we set the hidden value in ActivityIndicator.js?

Something like
```
hidden={this.props.hidesWhenStopped && !this.props.animating}
```

https://github.com/facebook/react-native/blob/master/Libraries/Components/ActivityIndicator/ActivityIndicator.js#L84

**Comment by redmar** [comment]
@janicduplessis have you tried this already? For me this doesn't seem to work, but maybe i'm doing something wrong. (used master, reproduced existing bug, and then applied your suggestion but i don't see any changes while i can confirm the code was really changed).

a possible reason for wanting to use the current patch (which now is only a couple of lines) is that it's an iOS bug only AFAIK so the fix also belongs to the iOS layer/space. But I'm pretty new to this repo so I'm not up to date yet with any policies about where to apply fixes or priority between JS or obj-c space. 😃

File: React/Views/RCTActivityIndicatorViewManager.m [hunk 1]
```diff
@@ -23,13 +23,30 @@ @implementation RCTConvert (UIActivityIndicatorView)
 
 @end
 
+@interface RCTActivityIndicatorView : UIActivityIndicatorView
```
Comment (janicduplessis) [inline React/Views/RCTActivityIndicatorViewManager.m]: Let's move this to it's own file for consistency with other views / view managers.

File: React/Views/RCTActivityIndicatorViewManager.m [hunk 2]
```diff
@@ -23,13 +23,30 @@ @implementation RCTConvert (UIActivityIndicatorView)
...
+@end
+
+@implementation RCTActivityIndicatorView {
+}
+
+- (void)setHidden:(BOOL)_hidden
```
Comment (janicduplessis) [inline React/Views/RCTActivityIndicatorViewManager.m]: Nit: Don't prefix local vars with an underscore.

**Comment by janicduplessis** [comment]
Ok let's do it this way. It doesn't fix the root issue of showing a view that is hidden initialy but I think this is so specific that activityindicator is the only control that this can cause an issue so fixing it here is fine.

**Comment by redmar** [comment]
@janicduplessis committed changes based on your comments, hopefully i've done it all in the correct way. (changing and seeing the diff from those .pbxproj files always feels weird)

**Comment by janicduplessis** [comment]
Thanks @redmar! Sorry for the delay, looks good, lets ship this.

@facebook-github-bot shipit

**Comment by facebook-github-bot-9** [comment]
Thanks for importing. If you are an FB employee go to [Phabricator](https://our.intern.facebook.com/intern/opensource/github/pull_request/1689230588064663/int_phab) to review.

**Comment by rainbowhat** [comment]
How do I get this fix? 
i'm on 
"dependencies": {
    "react": "15.1.0",
    "react-native": "^0.28.0"
  }

**Comment by redmar** [comment]
It's not yet in v0.29.0-rc.0 AFAIK so you have to wait on the next rc but you can use it now by using master temporarily in your package.json with the following:

```
"dependencies": {
  "react": "~15.2.0-rc.1",
  "react-native": "https://github.com/facebook/react-native.git"
}
```
Don't forget to run an ```react-native upgrade``` after doing ```npm install``` after changing your package.json 😄

**Comment by janicduplessis** [comment]
It should be in the final 0.29 release but it is not in the current 0.29rc.

**Pull request closed without merging**